### PR TITLE
NTP-667: Refine pricing on shortlist

### PR DESCRIPTION
**Pull request by kieranohanlon**
## Context

Allow the users to refine the pricing on their shortlist using group size and tuition type

## Changes proposed in this pull request

Add 2 new group size and tuition type drop down lists and when items are selected then refine the data that is shown in the shortlist results.

## Guidance to review

AC1: Tutor group sizes box (TOP) is selected then the data is updated to reflect this
GIVEN the compare tuition partners page has been displayed
WHEN the tutor group size has been selected
THEN the data in the shortlist table is automatically (need to use the button if no javascript) updated to reflect the user selection.  With prices reflecting the selected group size.  Any TPs not offering the group size are displayed at the bottom of the table with an appropriate message (e.g. Does not offer group sizes of 1 to 1).

AC2: Type of tuition box (TOP) is selected then the data is updated to reflect this
GIVEN the compare tuition partners page has been displayed
WHEN the type of tuition has been selected
THEN the data in the shortlist table is automatically (need to use the button if no javascript) updated to reflect the user selection.  With prices reflecting the selected type of tuition.  Any TPs not offering the type of tuition are displayed at the bottom of the table with an appropriate message (e.g. Does not offer in school tuition in Stockport).

AC3: Tutor group sizes AND type of tuition box (TOP) are selected
GIVEN the compare tuition partners page has been displayed
WHEN the group size and type of tuition has been selected
THEN the data in the shortlist table is automatically (need to use the button if no javascript) updated to reflect the user selections.  With prices reflecting the selected options.  Any TPs not offering these options are displayed at the bottom of the table with an appropriate message, matching AC1, AC2 or a new message if appropriate  (e.g. Does not offer group sizes of 1 to 1 or in school tuition in Stockport).

AC4: Can no longer sort the shortlist table by the Name column
GIVEN the compare tuition partners page has been displayed
THEN the table no longer allows the Name column to be sorted

AC5: The default sorting works as expected when the refinement filters are applied
GIVEN the compare tuition partners page has been displayed
WHEN the refinement filters are applied
THEN the TPs that match the refinement filters are displayed at the top of the table in the order that the TPs were added to the shortlist.  Those that do not match the refinement filters are shown at the bottom of the table in the order that these TPs were added to the shortlist.

AC6: The pricing sorting works as expected when the refinement filters are applied
GIVEN the compare tuition partners page has been displayed
WHEN the refinement filters are applied and the pricing is sorted
THEN the TPs that match the refinement filters are displayed at the top of the table in price order, either ascending or descending depending how many times the price column sort has been selected/sorted.  Those that do not match the refinement filters are shown at the bottom of the table in the order that these TPs were added to the shortlist.

AC7: Navigating away from the refined shortlist and returning retains the selections
GIVEN the compare tuition partners page has been displayed, the refinement options have been selected and the price sorting applied
WHEN navigating away from the page (e.g. going to the TP details or using the back link to the search results) and then returning to the shortlist (within the same session and using the website links)
THEN the previously selected refinement filters and sorting should be applied automatically

AC8: The shortlist tuition type refinement drop down will initially match the tuition type filter selected on the search results
GIVEN the user has selected a tuition type filter on the search results and have not been to the shortlist before
WHEN they go to the shortlist page
THEN the shortlist tuition type refinement drop down will initially match the tuition type filter selected on the search results.  If this tuition type refinement value is changed on the shortlist page then this will be retained going forward for this session and will not be kept in sync with the search results tuition type filter.

AC: The AC listed above should be tested with and without javascript enabled

## Link to Jira ticket

https://dfedigital.atlassian.net/browse/NTP-667

## Things to check

- [x] Title is in the format "NTP-XXX: Short description" where NTP-XXX is the Jira ticket reference
- [x] Code and tests follow the [coding standards](/docs/coding-standards.md)
- [x] `dotnet format` has been run in the repository root
- [x] Test coverage of new code is at least 80%
- [x] All UI related acceptance criteria specified in the ticket have been added to the relevant [cypress test feature file(s)](/UI/cypress/e2e/)
- [x] Unless explicitly mentioned in the ticket, all UI work should have been tested as working on desktop and mobile resolutions with JavaScript on and off
- [x] Database migrations can be applied to the current codebase in main without causing exceptions
- [x] Debug logging has been added after all logic decision points
- [x] All [automated testing](/README.md#testing) including accessibility and security has been run against your code
- [x] **PR deployment has been signed off by wider team**

### CODE REVIEW

File: Infrastructure/TuitionPartnerService.cs [hunk 1]
```diff
@@ -39,15 +115,22 @@ public IEnumerable<TuitionPartnerResult> OrderTuitionPartners(IEnumerable<Tuitio
...
+                        .OrderBy(e => e.Prices == null ? int.MaxValue : e.Prices!.Min(x => x.HourlyRate));
 
             case TuitionPartnerOrderBy.SeoList:
                 if (ordering.SeoUrlOrderBy == null || ordering.SeoUrlOrderBy.Length == 0)
                 {
                     return results;
                 }
-                return results.OrderBy(x => Array.IndexOf(ordering.SeoUrlOrderBy, x.SeoUrl));
+                return results
+                    .OrderBy(x => x.Prices == null ? (1000 + Array.IndexOf(ordering.SeoUrlOrderBy, x.SeoUrl)) : Array.IndexOf(ordering.SeoUrlOrderBy, x.SeoUrl));
```
Comment (hasan3d) [inline Infrastructure/TuitionPartnerService.cs:133]: Can we please extract the `1000` from here and declare a variable with a meaningful name?

File: Infrastructure/TuitionPartnerService.cs [hunk 2]
```diff
@@ -39,15 +115,22 @@ public IEnumerable<TuitionPartnerResult> OrderTuitionPartners(IEnumerable<Tuitio
 
             case TuitionPartnerOrderBy.MinPrice:
                 return ordering.Direction == OrderByDirection.Descending
-                    ? results.OrderBy(e => e.Name).OrderByDescending(e => e.Prices.Min(x => x.HourlyRate))
-                    : results.OrderBy(e => e.Name).OrderBy(e => e.Prices.Min(x => x.HourlyRate));
+                    ? results
+                        .OrderBy(e => e.Name)
+                        .OrderBy(s => (ordering.SeoUrlOrderBy == null || ordering.SeoUrlOrderBy.Length == 0) ? -1 : Array.IndexOf(ordering.SeoUrlOrderBy, s.SeoUrl))
```
Comment (hasan3d) [inline Infrastructure/TuitionPartnerService.cs:120]: Using multiple sequential `OrderBy` invocations in a query will overwrite the previous sort order, and only the last `OrderBy` will take effect. To sort by multiple criteria, you should use `OrderBy` followed by one or more `ThenBy` clauses. `ThenBy` is used to sort by additional criteria after the initial sort order has been established with `OrderBy`.

File: UI/Extensions/SearchModelExtensions.cs [hunk 3]
```diff
@@ -38,12 +27,56 @@ public static string ToQueryString(this SearchModel? model)
...
             this SearchModel model,
             Expression<Func<SearchModel, IEnumerable<T>?>> expression,
-            Dictionary<string, string> dictionary)
+            Dictionary<string, string> dictionary,
+            bool flattenCollection)
         {
-            model.Add(expression, (prop, name) => BuildQueryString(prop, name), dictionary);
+            if (flattenCollection)
+            {
+                model.Add(expression, (prop, name) => BuildQueryString(prop, name), dictionary);
```
Comment (hasan3d) [inline UI/Extensions/SearchModelExtensions.cs:56]: `model.Add(expression, (prop, name) => BuildQueryString(prop, name), dictionary);` can be converted into a method group by replacing the lambda expression with a method group, like so:
 `model.Add(expression, BuildQueryString, dictionary);` It is less code.

File: UI/Pages/Shortlist.cshtml [hunk 4]
```diff
@@ -21,13 +21,64 @@
...
             {
                 var dataQueryString = Model.Data.ToQueryString();
+
+                <input asp-for="Data.From" type="hidden" />
+                <input asp-for="Data.Postcode" type="hidden" />
+                @if (Model.Data.KeyStages != null)
+                {
+                    @for(int i = 0; i < Model.Data.KeyStages.Count(); i++)
+                    {
+                        <input asp-for="Data.KeyStages[i]" type="hidden" />
```
Comment (hasan3d) [inline UI/Pages/Shortlist.cshtml:31]: Dereference of a possibly null reference. I suggest the below change

` <input asp-for="Data.KeyStages![i]" type="hidden" />`

**Review by hasan3d — COMMENTED** [review]
The code looks good overall. There are a fair amount of code changes involved with this PR. However, I made some suggestions. If you can address these, that would be lovely. Thank you!

File: Infrastructure/TuitionPartnerService.cs (hunk 1 above)
Comment (kieranohanlon) [reply Infrastructure/TuitionPartnerService.cs]: Done

File: Infrastructure/TuitionPartnerService.cs (hunk 2 above)
Comment (kieranohanlon) [reply Infrastructure/TuitionPartnerService.cs]: Done for all the sorts, although this changes the "repeatable random" ordering that was used.

File: UI/Extensions/SearchModelExtensions.cs (hunk 3 above)
Comment (kieranohanlon) [reply UI/Extensions/SearchModelExtensions.cs]: Done

File: UI/Pages/Shortlist.cshtml (hunk 4 above)
Comment (kieranohanlon) [reply UI/Pages/Shortlist.cshtml]: Done - as mentioned, this wasn't flagging as an error in VS for some reason!

**Review by hasan3d — DISMISSED** [review]
lgtm

**Review by hasan3d — APPROVED** [review]
lgtm

**Pull request merged**